### ISSUE TITLE
Problem mit Umsatzcodierung / decodierung

### ISSUE THREAD

**Issue by FleiTec** [a-sit-plus/at-registrierkassen-mustercode#570]
Hallo.
Hab hier ein interessantes Phänomen beim verschlüsseln vom Umsatz.
Bisher waren alle Rechnungen in Ordnung.

Nun verschlüssle ich
Umsatz: 444756 CENT
Kassaid: 752
Rechnungsnummer: BAR.105

Beim entschlüsseln kommt raus:
Umsatz: 444755

Also wurde mir einfach ein Cent geklaut. Könnt ihr das nachvollziehen? 

Vielen Dank

**Comment by cmaxmedia** [comment]
Mein Tool liefert folgendes Ergebnis:

Umsatz Ein: 444756
Umsatz Encoded: 23/6Acb9dvrOw0cvrE8D8w==
Umsatz Decoded: 444756

In deinem Beispiel wird eine Ganzzahl geklaut - kann es sein, dass hier noch irgendwo mit Nachkommastellen gerechnet und gerundet wird...

**Comment by eauth** [comment]
Dazu müssten Sie etwas ausführlicher berichten wie die Verschlüsselung erfolgt. Es könnte z.B. sein dass der Umsatz als Gleitkommazahl vorliegt, da könnte genau so ein Phänomen passieren. Bei der eigentlichen Verschlüsselung/Entschlüsselung passiert der Fehler wohl kaum, wenn da irgendwas nicht stimmt würde eine komplett falsche Zahl herauskommen.

**Comment by FleiTec** [comment]
Stimmt- wenn ich 444756 direkt vor verschlüsselung nochmals hard eingebe- stimmt das Ergebnis.
Scheint da noch ein Problem zu geben!
VIELEN Dank für Eure Hilfe

**Comment by FleiTec** [comment]
Lösung bei PHP
Musste Var in string umwandeln:
$umsatz="$umsatz";

Jetzt gehts ;-)

**Comment by eauth** [comment]
Ich hatte das selbe Problem mit _$umsatz*100_, mit _(string)($umsatz*100)_ hats funktioniert. Ziemlich blöde Sache weil mit so einem Fehler das DEP im Eimer ist.

**Comment by lebail** [comment]
Nur zur Erklärung, _warum_ das so ist:

Die Ursache liegt darin, dass Sie zuerst mit Fließkommazahlen rechnen (die dank der Binärdarstellung nicht jeden Wert exakt repräsentieren können) und den Wert im Zuge der weiteren Verarbeitung dann irgendwann in den Typ `Integer` konvertieren (denn diesen benötigen Sie für die Verschlüsselung des Umsatzzählers).

Hat z.B. `$umsatz` den Double-Wert `19.99`, dann ergibt sich für `$umsatz * 100` der Double-Wert `1998.9999999999`.

Wenn Sie diesen Wert später nach Integer konvertieren, dann werden die Kommastellen abgeschnitten, somit entsteht der unerwartete Wert `1998`.

Konvertieren Sie den Wert `1998.9999999999` hingegen zuerst in einen String, dann wird er im Zuge dessen kaufmännisch gerundet, und es ergibt sich `1999`, was auch bei einer späteren Integer-Konversion erhalten bleibt.

Einen vergleichbaren Effekt könnten Sie statt mit `(string)` auch mit der `round()`-Funktion erzielen, was meiner Meinung nach sicherer wäre. Es ist nämlich nirgendwo garantiert, dass (string) _jeden_ beliebigen "fast ganzen" Wert auf eine Ganzzahl rundet - und ihn nicht doch unverändert übernimmt (womit das Problem eventuell bei manchen Zahlenwerten wieder auftauchen würde).

Außerdem brauchen Sie ja gar keinen String, es geht Ihnen vielmehr um das Runden der Fließkommazahl. Da ist die Umwandlung in einen String eher verwirrend - auch für jene, die vielleicht zukünftig an Ihrer Software weiterarbeiten und das (string) eventuell weglöschen, weil es auf den ersten Blick überflüssig ist.

**Comment by andreasschiestl** [comment]
Für Finanzanwendungen bei denen ein exaktes Ergebnis notwendig ist (wie in einer Kassensoftware) sollte von Gleitkomma-Datentypen ohnehin Abstand genommen werden. Unter Umständen könne sich diese Fehler durch Berechnungen aufsummieren, sodass auch bereits die zweite Stelle nach dem Komma betroffen ist. Da hilft dann auch kein Runden mehr.

**Comment by eauth** [comment]
Wenn man auf Nummer Sicher gehen möchte dann kann man in PHP die BCMath Funktionen verwenden, und ist damit auch unabhängig von den versions- und plattformabhängigen Integer/Floatingpoint Größen.
